### ISSUE TITLE
oQ: deepseek_v4 fails with "Missing mtp.0.{e,h}_proj.biases" after #TEMP guard — concrete repro + fix paths

### ISSUE THREAD

**Issue by alejandroed** [jundot/omlx#1296]
## Summary

`POST /admin/api/oq/start` with `model_path` pointing at the upstream `deepseek-ai/DeepSeek-V4-Flash` checkpoint and `preserve_mtp=true` fails immediately because of the `# TEMP` guard in [`omlx/oq.py:2222`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/oq.py#L2222). After commenting out the guard locally to see what would happen next, the run gets further but fails during sensitivity with a concrete missing-parameters error. Filing this so the follow-up the TODO mentions has a precise reproduction + a partial diagnosis attached.

## Environment

- oMLX `v0.3.9.dev2` (commit `9749c40`) from `git checkout`
- mlx-lm `0.31.3`, mlx `0.31.2`, mlx-audio `0.4.3`
- Apple Silicon Mac Studio, macOS 26.2, 512 GB unified memory, ~580 GB free disk
- Model: `deepseek-ai/DeepSeek-V4-Flash` (full FP8/FP4-mixed, 160 GB, snapshot `6976c7ff1b30a1b2cb7805021b8ba4684041f136`)
- No other models pinned/loaded while oQ ran (RAM was free)

## Reproduction

1. Download the upstream model:
   ```bash
   HF_HUB_ENABLE_HF_TRANSFER=1 hf download deepseek-ai/DeepSeek-V4-Flash
   ```
2. Symlink it under the oMLX model dir:
   ```bash
   mkdir -p ~/.omlx/models/deepseek-ai
   ln -sfn ~/.cache/huggingface/hub/models--deepseek-ai--DeepSeek-V4-Flash/snapshots/6976c7ff1b30a1b2cb7805021b8ba4684041f136 ~/.omlx/models/deepseek-ai/DeepSeek-V4-Flash
   ```
3. Comment out the guard in `omlx/oq.py:2222` (the `if config.get("model_type") == "deepseek_v4": raise RuntimeError(...)` block).
4. Restart the server and unload any pinned models so RAM is free.
5. Call:
   ```bash
   curl -X POST -H "Authorization: Bearer $KEY" -H "Content-Type: application/json" \
     http://localhost:54330/admin/api/oq/start \
     -d '{
       "model_path": "/Users/systems/.omlx/models/deepseek-ai/DeepSeek-V4-Flash",
       "oq_level": 4,
       "dtype": "bfloat16",
       "preserve_mtp": true,
       "text_only": true
     }'
   ```

## Observed

The task progresses through FP8 dequant + eager sanitize + non-quantizable module discovery (the 43 `model.layers.*.ffn.gate` modules, expected). It then fails in `_measure_sensitivity`:

```
2026-05-18 19:17:57,910 - omlx.oq - INFO - oQ4: measuring layer sensitivity for streaming path
2026-05-18 19:17:58,419 - omlx.oq - ERROR - Sensitivity measurement: model load failed (Missing 2 parameters:
mtp.0.e_proj.biases,
mtp.0.h_proj.biases.)
2026-05-18 19:17:58,457 - omlx.admin.oq_manager - ERROR - oQ quantization failed: DeepSeek-V4-Flash -> oQ4:
sensitivity measurement produced no scores. Check the preceding log lines for the root cause (model load,
calibration data, or layer discovery), and either fix it or pass an explicit sensitivity_model_path.
```

i.e. exactly the failure mode the `# TEMP` guard predicts for the raw-load path.

## Root cause (best guess after walking the code)

`MTPBlock.e_proj` / `h_proj` are declared as `nn.Linear(dim, dim, bias=False)` in [`omlx/patches/mlx_lm_mtp/deepseek_v4_model.py:135-136`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/patches/mlx_lm_mtp/deepseek_v4_model.py#L135-L136). The upstream DSV4 checkpoint ships those tensors in **FP8 (`.weight + .scale`)**. When `model.load_weights()` sees `.scale` it promotes the layer to `QuantizedLinear`, which then mandates a `.biases` tensor that the checkpoint does not carry.

The MTP-aware sanitize at [`omlx/patches/mlx_lm_mtp/deepseek_v4_model.py:351`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/patches/mlx_lm_mtp/deepseek_v4_model.py#L351) does preserve `mtp.*` keys when `has_mtp_weights` is true, but the comment in `_measure_sensitivity` ([`omlx/oq.py:3060-3068`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/oq.py#L3060-L3068) ish) talks about the VLM path (`language_model.mtp.*`) — it isn't obvious that the MTP runtime patch is being applied on the LLM-only DSV4 path during sensitivity load.

The base sanitize at [`omlx/patches/deepseek_v4/deepseek_v4_model.py:1042-1047`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/patches/deepseek_v4/deepseek_v4_model.py#L1042-L1047) unconditionally drops `mtp.*`:

```python
for k, v in weights.items():
    if k.startswith("mtp."):
        continue
    ...
```

so if the MTP-aware sanitize isn't installed on the model class for the sensitivity load, `mtp.*` gets stripped → load reports `mtp.0.{e,h}_proj.biases` missing (the model class itself still expects them).

## Candidate fixes (didn't apply locally — out of confidence on shapes)

1. Apply `apply_mlx_lm_mtp_patch` in the LLM branch of `_measure_sensitivity` (it currently looks installed for VLM only) so the MTP-aware sanitize replaces the stripping one.
2. In the MTP-aware sanitize, dequantize FP8 `e_proj`/`h_proj` (`.weight + .scale`) to bf16 with `.scale` dropped, so `nn.Linear` stays a plain `nn.Linear` and no `.biases` is needed.
3. Synthesize zero `.biases` of the expected shape for those two tensors when `preserve_mtp=true`. Skipped this locally because I'd be guessing the zero-point convention and could silently mis-quantize.

I'm happy to take a shot at (2) if you can sanity-check that's the direction you'd accept — it feels like the smallest change with the lowest blast radius.

## Heads-up about the second guard predicate

The `# TEMP` guard also warns about `_forward_layer` not matching `DeepseekV4Block.__call__(h, mask, cache, input_ids)`. We didn't get past the missing-biases error to confirm this manifests, but skimming `_measure_sensitivity_from_model` at [`omlx/oq.py:3022`](https://github.com/jundot/omlx/blob/v0.3.9.dev2/omlx/oq.py#L3022) it does call `_forward_layer(block, inputs, layer_mask, position_ids)`, and `DeepseekV4Block` needs `input_ids` (not `position_ids`) for MoE expert routing in `self.ffn(self.ffn_norm(x), input_ids)`. So even with the missing-biases fix, sensitivity per-layer forward would either silently route experts wrong or hit the all-signatures-failed branch.

## Why this matters

`mlx-community` shipping no MLX variant with `mtp.*` preserved (every existing conversion goes through HF transformers which drops them at load time), `LordNeel/DeepSeek-V4-Flash-Acti-MTP-W4A16-FP8` being vLLM/CUDA-only, and the Blaizzy PR being draft together mean **oQ with `preserve_mtp=true` is currently the only viable path to an MLX-compatible V4 + MTP build**. Worth unblocking when it fits the roadmap.

Thanks for the work on 0.3.9 in general — happy to test patches.

**Comment by jundot** [comment]
Thanks for the detailed repro. The `#TEMP` guard is intentional. DeepSeek oQ isn't supported on 0.3.9 yet, and the missing-bias error you hit after bypassing the guard is exactly the path it's meant to block. For now please use the official MLX Quant releases on mlx-community. I'll announce here when DeepSeek oQ support is ready.
